iceoryx2-bb/cxx/tests/src/optional_tests.cpp

```diff
@@ -0,0 +1,1008 @@
+    ASSERT_EQ(Observable::s_counter.wasMoveConstructed, 0);
+    ASSERT_EQ(Observable::s_counter.wasMoveAssigned, 0);
+}
+TEST_F(OptionalFixture, rvalue_value_or_with_rvalue_argument_moves_fallback_on_empty_optional) {
+    int const fallback_tracking_id = 225;
+    iox2::container::Optional<Observable> sut;
+    Observable fallback;
+    fallback.id = fallback_tracking_id;
+    Observable::s_counter.wasInitialized = 0;
+    Observable::s_counter.wasCopyConstructed = 0;
+    Observable::s_counter.wasCopyAssigned = 0;
+    Observable::s_counter.wasMoveConstructed = 0;
+    Observable::s_counter.wasMoveAssigned = 0;
+    ASSERT_EQ(std::move(sut).value_or(std::move(fallback)).id, fallback_tracking_id);
+    ASSERT_EQ(Observable::s_counter.wasInitialized, 0);
+    ASSERT_EQ(Observable::s_counter.wasCopyConstructed, 0);
+    ASSERT_EQ(Observable::s_counter.wasCopyAssigned, 0);
+    ASSERT_EQ(Observable::s_counter.wasMoveConstructed, 1);
+    ASSERT_EQ(Observable::s_counter.wasMoveAssigned, 0);
+}
+
+TEST(Optional, reset_leaves_empty_optional_in_empty_state) {
+    iox2::container::Optional<int> sut;
+    ASSERT_TRUE(!sut.has_value());
+    sut.reset();
+    ASSERT_TRUE(!sut.has_value());
+}
+
+TEST(Optional, reset_puts_full_optional_to_empty_state) {
+    int const contained_value = 42;
+    iox2::container::Optional<int> sut { contained_value };
+    ASSERT_TRUE(sut.has_value());
+    sut.reset();
+    ASSERT_TRUE(!sut.has_value());
+}
+
+TEST_F(OptionalFixture, reset_on_full_optional_destructs_contained_value) {
+    {
+        iox2::container::Optional<Observable> sut { Observable {} };
+        ASSERT_TRUE(sut.has_value());
+        Observable::s_counter.wasDestructed = 0;
+        sut.reset();
+        ASSERT_TRUE(!sut.has_value());
+        ASSERT_EQ(Observable::s_counter.wasDestructed, 1);
+        Observable::s_counter.wasDestructed = 0;
+    }
+    ASSERT_EQ(Observable::s_counter.wasDestructed, 0);
+}
+
+TEST(Optional, operator_arrow_should_bypass_overloaded_operator_ampersand) {
+    int const tracking_id = 54321;
+    iox2::container::testing::CustomAddressOperator obj;
+    obj.id = tracking_id;
+    iox2::container::Optional<iox2::container::testing::CustomAddressOperator> sut { obj };
+    iox2::container::testing::CustomAddressOperator::s_count_address_operator = 0;
+    ASSERT_EQ(sut->id, tracking_id);
+    ASSERT_EQ(iox2::container::testing::CustomAddressOperator::s_count_address_operator, 0);
+}
+
+TEST(Optional, const_operator_arrow_should_bypass_overloaded_operator_ampersand) {
+    int const tracking_id = 54321;
+    iox2::container::testing::CustomAddressOperator obj;
+    obj.id = tracking_id;
+    iox2::container::Optional<iox2::container::testing::CustomAddressOperator> const sut { obj };
+    iox2::container::testing::CustomAddressOperator::s_count_address_operator = 0;
+    ASSERT_EQ(sut->id, tracking_id);
+    ASSERT_EQ(iox2::container::testing::CustomAddressOperator::s_count_address_operator, 0);
+}
+
```

Comment:
Maybe I missed it and it's also kind of a philosophical question. Should there be an explicit test that verifies that  a moved-from Optional is empty?

This would require to deactivate some warnings for the scope of the test.

Reply:
Point 2 will be difficult to accomplish. Depending on who is using it, it might be surprising that it does not behave like the unsafe STL. But I get what you mean.

Reply:
The problem is, if the use case is drop-in replacability for `std::optional`, nulling on move does not necesarily make it safer. Consider:

```cpp
std::optional<std::vector<int>> opt{ std::vector<int>{ 1, 2, 3, 4, 5 } };
std::vector<int> moved_out = std::move(opt).value();
*opt = std::vector<int>{ 6, 7, 8, 9 };
assert(opt->at(3) == 9);
```

This is valid for C++17 optional but would trigger language UB for an empties-on-move optional. Code that was written for C++17 `optional` would not be able to migrate to iox2's `optional` safely.

Reply:
Tracked as #1001.

Here is a variation of the example above that is not detected by clang-tidy and is not detectable at runtime by a check that would terminate the program:

```cpp
std::optional<std::vector<int>> opt{ std::vector<int>{ 1, 2, 3, 4, 5 } };
std::optional<std::vector<int>>& moved_from = opt;
std::vector<int>& the_element = *opt;
std::vector<int> moved_out = std::move(moved_from).value();
*opt = std::vector<int>{ 6, 7, 8, 9 };
assert(the_element->at(3) == 9);
```

Breaking this leads to, what I would argue is, the worst kind of UB, as it is very likely to silently work on lower optimization levels, but also rather likely to break randomly on higher optimization levels.